db/commitlog/commitlog.cc

```diff
@@ -1230,6 +1278,119 @@ db::commitlog::segment_manager::segment_manager(config c)
     }
 }
 
+void db::commitlog::segment_manager::send(message_payload m, message_clock::time_point t) {
+    _message_queue.emplace_back(message{ t, std::move(m) });
+    std::push_heap(_message_queue.begin(), _message_queue.end(), mcmp);
+    auto nt = _message_queue.front().timeout;
+    if (_message_loop_timer.armed() && nt < _message_loop_timer.get_timeout()) {
+        auto now = message_clock::now();
+        if (nt < now) {
+            _message_loop_timer.cancel();
+            wakeup();
+        } else {
+            _message_loop_timer.rearm(nt);
+        }
+    }
+}
+
+void db::commitlog::segment_manager::wakeup() {
+    if (_message_loop_wakeup) {
+        _message_loop_wakeup->set_value();
+        _message_loop_wakeup = nullptr;
+    }
+    ++_message_loop_wakeups;
+}
+
+struct db::commitlog::segment_manager::message_wait {
+    segment_manager& _manager;
+    promise<> _promise;
+
+    message_wait(segment_manager& m)
+        : _manager(m)
+    {
+        assert(_manager._message_loop_wakeup == nullptr);
+        _manager._message_loop_wakeup = &_promise;
+    }
+    ~message_wait() {
+        _manager._message_loop_wakeup = nullptr;
+    }
+    future<> get_future() {
+        return _promise.get_future();
+    }
+};
+
+class db::commitlog::segment_manager::message_partition {
```

Comment:
What is a message_partition?

Reply:
A partition of a vector of messages, by message type. So we can neatly iterate over only the messages we care about. A wrapper around std::partition + erasing the partitioned messages when we're done.